File: go/schemapb/compute.go
package schemapb

import (
	"fmt"
	"strconv"
	"time"
)

// Resolve fills unset fields from their defaults, coerces string inputs (when
// the schema enables coercion), applies normalize expressions and evaluates
// Computed fields in dependency order. values is mutated in place and
// returned; the result reports expression failures (code EXPR_ERROR). A nil
// map starts an empty form.
//
// Value model: native Go (see value.go). Callers holding a wire StructValue
// convert with ToGo()/StructFromGo().
func (s *Schema) Resolve(values map[string]any) (map[string]any, *ValidationResult, error) {
	e, err := s.engine()
	if err != nil {
		return values, nil, err
	}
	out, res := e.Resolve(values)
	return out, res, nil
}

// Resolve is the compiled-engine form of (*Schema).Resolve.
func (e *Engine) Resolve(values map[string]any) (map[string]any, *ValidationResult) {
	if values == nil {
		values = map[string]any{}
	}
	res := &ValidationResult{}
	var tasks []computeTask
	e.seed(e.schema, values, "", &tasks, values, res)
	e.runNormalize(e.schema, values, values, res)
	e.runCompute(values, tasks, res)
	return values, res
}

// exprErr builds a runtime expression-failure ValidationError.
func exprErr(path, expr, msg string) *ValidationError {
	return &ValidationError{
		Path:     path,
		Code:     ErrorCode_ERROR_CODE_EXPR_ERROR,
		Expr:     expr,
		Severity: SeverityError,
		Message:  msg,
	}
}

// active evaluates a field's `when` gate over root. A field with no gate is
// always active. An evaluation error deactivates the field and is reported.
func (e *Engine) active(f *Schema_Field, root map[string]any, path string, res *ValidationResult) bool {
	when := f.GetWhen()
	if when == "" {
		return true
	}
	ok, err := e.evalBool(when, map[string]any{"this": nil, "root": root})
	if err != nil {
		res.Errors = append(res.Errors, exprErr(path, when, "when: "+err.Error()))
		return false
	}
	return ok
}

// computeTask is a Computed field pending evaluation: it reads from root
// (shared) and writes its result to scope[field.Name].
type computeTask struct {
	field *Schema_Field
	scope map[string]any
	path  string
}

// seed fills defaults for unset fields (immutable fields are forced to their
// default) and collects Computed fields as tasks, recursing into present
// containers. It never materialises an absent object, so optional sub-forms
// stay absent and don't spuriously trip their children's "required" checks.
func (e *Engine) seed(schema *Schema, scope map[string]any, prefix string, tasks *[]computeTask, root map[string]any, res *ValidationResult) {
	coerce := schema.GetCoerce()
	for _, f := range schema.GetFields() {
		name := f.GetName()
		path := joinPath(prefix, name)

		// Inactive fields are treated as absent: no default seeded, no
		// Computed scheduled, subtree not recursed. Existing value preserved.
		if !e.active(f, root, path, res) {
			continue
		}

		if coerce {
			if cur, ok := scope[name]; ok {
				if coerced, changed := coerceInput(f, cur); changed {
					scope[name] = coerced
				}
			}
		}

		if f.GetImmutable() {
			if dv, ok := defaultValue(f); ok {
				scope[name] = dv
			}
		} else if _, ok := scope[name]; !ok {
			if dv, ok := defaultValue(f); ok {
				scope[name] = dv
			}
		}

		switch {
		case f.GetComputed() != nil:
			*tasks = append(*tasks, computeTask{field: f, scope: scope, path: path})
		case f.GetObject() != nil && f.GetObject().GetSchema() != nil:
			if child, ok := scope[name].(map[string]any); ok {
				e.seed(f.GetObject().GetSchema(), child, path, tasks, root, res)
			}
		case f.GetList() != nil && len(f.GetList().GetItems()) >= 1:
			if o := f.GetList().GetItems()[0].GetObject(); o != nil && o.GetSchema() != nil {
				if arr, ok := scope[name].([]any); ok {
					for i, el := range arr {
						if m, ok := el.(map[string]any); ok {
							e.seed(o.GetSchema(), m, fmt.Sprintf("%s[%d]", path, i), tasks, root, res)
						}
					}
				}
			}
		case f.GetMap() != nil && f.GetMap().GetValueSchema() != nil:
			if mm, ok := scope[name].(map[string]any); ok {
				for k, el := range mm {
					if m, ok := el.(map[string]any); ok {
						e.seed(f.GetMap().GetValueSchema(), m, joinPath(path, k), tasks, root, res)
					}
				}
			}
		case f.GetOneOf() != nil:
			if variant, m := selectVariant(f.GetOneOf(), scope[name]); variant != nil {
				e.seed(variant, m, path, tasks, root, res)
			}
		case f.GetRef() != nil:
			if def := e.schema.GetDefs()[refDefKey(f.GetRef())]; def != nil {
				if child, ok := scope[name].(map[string]any); ok {
					e.seed(def, child, path, tasks, root, res)
				}
			}
		}
	}
}

// selectVariant picks the OneOf variant schema for a value by its
// discriminator property; nil when the value is not an object or the
// discriminator is missing/unknown.
func selectVariant(oo *Schema_Field_OneOf, val any) (*Schema, map[string]any) {
	m, ok := val.(map[string]any)
	if !ok {
		return nil, nil
	}
	disc, ok := m[oo.GetDiscriminator()].(string)
	if !ok || disc == "" {
		return nil, nil
	}
	variant, ok := oo.GetVariants()[disc]
	if !ok {
		return nil, nil
	}
	return variant, m
}

// runNormalize applies normalize expressions to present, active fields,
// recursing into containers. Runs after seed (defaults in place) and before
// runCompute (Computed reads normalised values).
func (e *Engine) runNormalize(schema *Schema, scope, root map[string]any, res *ValidationResult) {
	for _, f := range schema.GetFields() {
		name := f.GetName()
		cur, exists := scope[name]
		if !exists || cur == nil {
			continue
		}
		if f.GetWhen() != "" {
			if ok, err := e.evalBool(f.GetWhen(), map[string]any{"this": nil, "root": root}); err != nil || !ok {
				continue
			}
		}
		if norm := f.GetNormalize(); norm != "" {
			out, err := e.eval(norm, map[string]any{"this": cur, "root": root})
			if err != nil {
				res.Errors = append(res.Errors, exprErr(name, norm, "normalize: "+err.Error()))
			} else {
				scope[name] = out
				cur = out
			}
		}
		if o := f.GetObject(); o != nil && o.GetSchema() != nil {
			if child, ok := cur.(map[string]any); ok {
				e.runNormalize(o.GetSchema(), child, root, res)
			}
		}
		if l := f.GetList(); l != nil && len(l.GetItems()) >= 1 {
			if obj := l.GetItems()[0].GetObject(); obj != nil && obj.GetSchema() != nil {
				if arr, ok := cur.([]any); ok {
					for _, el := range arr {
						if m, ok := el.(map[string]any); ok {
							e.runNormalize(obj.GetSchema(), m, root, res)
						}
					}
				}
			}
		}
		if mp := f.GetMap(); mp != nil && mp.GetValueSchema() != nil {
			if mm, ok := cur.(map[string]any); ok {
				for _, el := range mm {
					if m, ok := el.(map[string]any); ok {
						e.runNormalize(mp.GetValueSchema(), m, root, res)
					}
				}
			}
		}
		if oo := f.GetOneOf(); oo != nil {
			if variant, m := selectVariant(oo, cur); variant != nil {
				e.runNormalize(variant, m, root, res)
			}
		}
		if ref := f.GetRef(); ref != nil {
			if def := e.schema.GetDefs()[refDefKey(ref)]; def != nil {
				if m, ok := cur.(map[string]any); ok {
					e.runNormalize(def, m, root, res)
				}
			}
		}
	}
}

// runCompute evaluates tasks in dependency order (the root paths each
// expression reads), writing each result into its scope. Cycles between
// nested scopes are reported and their fields left unevaluated (top-level
// cycles are already rejected at Compile).
func (e *Engine) runCompute(root map[string]any, tasks []computeTask, res *ValidationResult) {
	if len(tasks) == 0 {
		return
	}
	byPath := make(map[string]computeTask, len(tasks))
	for _, t := range tasks {
		byPath[t.path] = t
	}
	deps := map[string][]string{}
	for _, t := range tasks {
		for _, d := range e.exprDeps(t.field.GetComputed().GetExpr()) {
			if d != t.path {
				if _, ok := byPath[d]; ok {
					deps[t.path] = append(deps[t.path], d)
				}
			}
		}
	}

	const (
		white = iota
		gray
		black
	)
	color := map[string]int{}
	var order []computeTask
	var visit func(string) bool
	visit = func(p string) bool {
		switch color[p] {
		case gray:
			return false
		case black:
			return true
		}
		color[p] = gray
		for _, d := range deps[p] {
			if !visit(d) {
				return false
			}
		}
		color[p] = black
		order = append(order, byPath[p])
		return true
	}
	for _, t := range tasks {
		if color[t.path] == black {
			continue
		}
		if !visit(t.path) {
			res.Errors = append(res.Errors, schemaErr(t.path, "computed field cycle"))
		}
	}

	for _, t := range order {
		c := t.field.GetComputed()
		out, err := e.eval(c.GetExpr(), map[string]any{"this": nil, "root": root})
		if err != nil {
			res.Errors = append(res.Errors, exprErr(t.path, c.GetExpr(), "compute: "+err.Error()))
			continue
		}
		shaped, err := shapeResult(c.GetResult(), out)
		if err != nil {
			res.Errors = append(res.Errors, exprErr(t.path, c.GetExpr(), "compute: "+err.Error()))
			continue
		}
		t.scope[t.field.GetName()] = shaped
	}
}

// shapeResult converts a computed result to the native representation of its
// declared ResultType. UNSPECIFIED keeps the CEL-native value as is (numbers
// stay honestly typed — no float64 flattening).
func shapeResult(rt Schema_Field_ResultType, x any) (any, error) {
	if x == nil {
		return nil, nil
	}
	switch rt {
	case Schema_Field_RESULT_TYPE_UNSPECIFIED, Schema_Field_RESULT_TYPE_JSON:
		return x, nil
	case Schema_Field_RESULT_TYPE_DOUBLE:
		if n, ok := asFloat64(x); ok {
			return n, nil
		}
	case Schema_Field_RESULT_TYPE_INT64:
		if n, ok := asInt64(x); ok {
			return n, nil
		}
	case Schema_Field_RESULT_TYPE_UINT64:
		if n, ok := asUint64(x); ok {
			return n, nil
		}
	case Schema_Field_RESULT_TYPE_BOOL:
		if b, ok := x.(bool); ok {
			return b, nil
		}
	case Schema_Field_RESULT_TYPE_STRING:
		if s, ok := x.(string); ok {
			return s, nil
		}
	case Schema_Field_RESULT_TYPE_DURATION:
		if d, ok := x.(time.Duration); ok {
			return d, nil
		}
	case Schema_Field_RESULT_TYPE_TIMESTAMP:
		if t, ok := x.(time.Time); ok {
			return t, nil
		}
	case Schema_Field_RESULT_TYPE_BYTES:
		if b, ok := x.([]byte); ok {
			return b, nil
		}
	}
	return nil, fmt.Errorf("result %T does not match declared type %v", x, rt)
}

// coerceInput converts a string value to the field's expected native type.
// Returns the coerced value and whether conversion occurred; unparseable
// strings pass through so the type error reports normally.
func coerceInput(f *Schema_Field, val any) (any, bool) {
	s, ok := val.(string)
	if !ok {
		return val, false
	}
	switch {
	case f.GetInt32() != nil, f.GetInt64() != nil:
		if n, err := strconv.ParseInt(s, 10, 64); err == nil {
			return n, true
		}
	case f.GetUint32() != nil, f.GetUint64() != nil:
		if n, err := strconv.ParseUint(s, 10, 64); err == nil {
			return n, true
		}
	case f.GetFloat() != nil, f.GetDouble() != nil:
		if n, err := strconv.ParseFloat(s, 64); err == nil {
			return n, true
		}
	case f.GetBool() != nil:
		switch s {
		case "true":
			return true, true
		case "false":
			return false, true
		}
	case f.GetEnum() != nil:
		if n, err := strconv.ParseInt(s, 10, 32); err == nil {
			return n, true
		}
	case f.GetDuration() != nil:
		if d, err := time.ParseDuration(s); err == nil {
			return d, true
		}
	case f.GetTimestamp() != nil:
		if t, err := time.Parse(time.RFC3339, s); err == nil {
			return t, true
		}
	}
	return val, false
}

// defaultValue returns a field's default in the native value model and
// whether one is set.
func defaultValue(f *Schema_Field) (any, bool) {
	switch {
	case f.GetFloat() != nil:
		if d := f.GetFloat().Default; d != nil {
			return float64(*d), true
		}
	case f.GetDouble() != nil:
		if d := f.GetDouble().Default; d != nil {
			return *d, true
		}
	case f.GetInt32() != nil:
		if d := f.GetInt32().Default; d != nil {
			return int64(*d), true
		}
	case f.GetInt64() != nil:
		if d := f.GetInt64().Default; d != nil {
			return *d, true
		}
	case f.GetUint32() != nil:
		if d := f.GetUint32().Default; d != nil {
			return uint64(*d), true
		}
	case f.GetUint64() != nil:
		if d := f.GetUint64().Default; d != nil {
			return *d, true
		}
	case f.GetBool() != nil:
		if d := f.GetBool().Default; d != nil {
			return *d, true
		}
	case f.GetString_() != nil:
		if d := f.GetString_().Default; d != nil {
			return *d, true
		}
	case f.GetBytes() != nil:
		if d := f.GetBytes().GetDefault(); d != nil {
			return append([]byte(nil), d...), true
		}
	case f.GetEnum() != nil:
		if d := f.GetEnum().Default; d != nil {
			return int64(*d), true
		}
	case f.GetDuration() != nil:
		if d := f.GetDuration().GetDefault(); d != nil {
			return d.AsDuration(), true
		}
	case f.GetTimestamp() != nil:
		if d := f.GetTimestamp().GetDefault(); d != nil {
			return d.AsTime(), true
		}
	case f.GetJson() != nil:
		if d := f.GetJson().GetDefault(); d != nil {
			return d.ToGo(), true
		}
	}
	return nil, false
}

// =============================================================================
// Renderer helpers
// =============================================================================

// FieldActive reports whether the named top-level field is active for the
// given form (its `when` gate over root). A field with no `when` is always
// active.
func (s *Schema) FieldActive(name FieldName, root map[string]any) (bool, error) {
	e, err := s.engine()
	if err != nil {
		return false, err
	}
	return e.FieldActive(name, root)
}

// FieldActive is the compiled-engine form of (*Schema).FieldActive.
func (e *Engine) FieldActive(name FieldName, root map[string]any) (bool, error) {
	f := findField(e.schema.GetFields(), string(name))
	if f == nil {
		return false, fmt.Errorf("schemapb: unknown field %q", name)
	}
	if f.GetWhen() == "" {
		return true, nil
	}
	return e.evalBool(f.GetWhen(), map[string]any{"this": nil, "root": root})
}

// EnumOptions returns the allowed integer values for the named top-level enum
// field given the form: the options_expr result when set, the static enum
// values otherwise.
func (s *Schema) EnumOptions(name FieldName, root map[string]any) ([]int32, error) {
	e, err := s.engine()
	if err != nil {
		return nil, err
	}
	return e.EnumOptions(name, root)
}

// EnumOptions is the compiled-engine form of (*Schema).EnumOptions.
func (e *Engine) EnumOptions(name FieldName, root map[string]any) ([]int32, error) {
	f := findField(e.schema.GetFields(), string(name))
	if f == nil || f.GetEnum() == nil {
		return nil, fmt.Errorf("schemapb: field %q is not an enum", name)
	}
	en := f.GetEnum()
	if en.GetOptionsExpr() == "" {
		out := make([]int32, 0, len(en.GetValues()))
		for k := range en.GetValues() {
			out = append(out, k)
		}
		return out, nil
	}
	return e.evalEnumOptions(en.GetOptionsExpr(), root)
}

// evalEnumOptions evaluates an options_expr into the allowed enum values.
func (e *Engine) evalEnumOptions(src string, root map[string]any) ([]int32, error) {
	res, err := e.eval(src, map[string]any{"this": nil, "root": root})
	if err != nil {
		return nil, err
	}
	arr, ok := res.([]any)
	if !ok {
		return nil, fmt.Errorf("options_expr yields %T, want a list of ints", res)
	}
	out := make([]int32, 0, len(arr))
	for _, el := range arr {
		n, ok := asInt64(el)
		if !ok {
			return nil, fmt.Errorf("options_expr element %v is not an int", el)
		}
		out = append(out, int32(n))
	}
	return out, nil
}

// ListCount returns the required length of the named top-level list field as
// derived from its count_expr over root.
func (s *Schema) ListCount(name FieldName, root map[string]any) (int64, error) {
	e, err := s.engine()
	if err != nil {
		return 0, err
	}
	return e.ListCount(name, root)
}

// ListCount is the compiled-engine form of (*Schema).ListCount.
func (e *Engine) ListCount(name FieldName, root map[string]any) (int64, error) {
	f := findField(e.schema.GetFields(), string(name))
	if f == nil || f.GetList() == nil {
		return 0, fmt.Errorf("schemapb: field %q is not a list", name)
	}
	ce := f.GetList().GetCountExpr()
	if ce == "" {
		return 0, fmt.Errorf("schemapb: field %q has no count_expr", name)
	}
	return e.evalCount(ce, root)
}

// evalCount evaluates a count_expr into a non-negative length.
func (e *Engine) evalCount(src string, root map[string]any) (int64, error) {
	res, err := e.eval(src, map[string]any{"this": nil, "root": root})
	if err != nil {
		return 0, err
	}
	n, ok := asInt64(res)
	if !ok || n < 0 {
		return 0, fmt.Errorf("count_expr yields %v, want a non-negative int", res)
	}
	return n, nil
}

// findField returns the field with the given name, or nil.
func findField(fields []*Schema_Field, name string) *Schema_Field {
	for _, f := range fields {
		if f.GetName() == name {
			return f
		}
	}
	return nil
}

File: py/src/schemapb/compute.py
"""The resolve pipeline, mirroring the Go reference compute.go."""

from __future__ import annotations

import base64
import binascii
import datetime as dt
from typing import TYPE_CHECKING, cast

from schemapb._gen.schemapb import (
    ErrorCode,
    Schema,
    SchemaField,
    SchemaFieldList,
    SchemaFieldOneOf,
    SchemaFieldRef,
    SchemaFieldResultType,
    SchemaFieldSeverity,
    ValidationError,
)
from schemapb.descriptor import join_path, schema_err
from schemapb.duration import parse_go_duration, parse_rfc3339
from schemapb.value import (
    Native,
    NativeStruct,
    as_float,
    as_int,
    as_uint,
    to_native,
)

if TYPE_CHECKING:
    from schemapb.engine import Engine


def expr_err(path: str, expr: str, msg: str) -> ValidationError:
    return ValidationError(
        path=path,
        code=ErrorCode.EXPR_ERROR,
        expr=expr,
        severity=SchemaFieldSeverity.ERROR,
        message=msg,
    )


def ref_def_key(ref: SchemaFieldRef) -> str:
    if ref.id is not None:
        return f"{ref.id.namespace}\x00{ref.id.name}\x00{ref.id.version}"
    return ref.name or ""


def is_tuple(l: SchemaFieldList) -> bool:  # noqa: E741
    return len(l.items) > 1


def list_item_def(l: SchemaFieldList, i: int) -> SchemaField | None:  # noqa: E741
    if len(l.items) == 1:
        return l.items[0]
    return l.items[i] if i < len(l.items) else None


def select_variant(
    oo: SchemaFieldOneOf,
    val: Native,
) -> tuple[Schema, NativeStruct] | None:
    if not isinstance(val, dict):
        return None
    disc = val.get(oo.discriminator)
    if not isinstance(disc, str) or disc == "":
        return None
    variant = oo.variants.get(disc)
    return None if variant is None else (variant, val)


def resolve(e: Engine, values: NativeStruct) -> list[ValidationError]:
    """Defaults, coercion, normalize, computed — in place."""
    errs: list[ValidationError] = []
    tasks: list[tuple[SchemaField, NativeStruct, str]] = []
    _seed(e, e.schema, values, "", tasks, values, errs)
    _run_normalize(e, e.schema, values, values, errs)
    _run_compute(e, values, tasks, errs)
    return errs


def field_is_active(
    e: Engine,
    f: SchemaField,
    root: NativeStruct,
    path: str,
    errs: list[ValidationError] | None,
) -> bool:
    when = f.when or ""
    if when == "":
        return True
    ok, err = e.eval_bool(when, {"root": root})
    if err is not None:
        if errs is not None:
            errs.append(expr_err(path, when, f"when: {err}"))
        return False
    return ok


def _seed(
    e: Engine,
    schema: Schema,
    scope: NativeStruct,
    prefix: str,
    tasks: list[tuple[SchemaField, NativeStruct, str]],
    root: NativeStruct,
    errs: list[ValidationError],
) -> None:
    coerce = schema.coerce
    for f in schema.fields:
        name = f.name
        path = join_path(prefix, name)
        if not field_is_active(e, f, root, path, errs):
            continue
        if coerce and name in scope:
            coerced = coerce_input(f, scope[name])
            if coerced is not None:
                scope[name] = coerced
        if f.immutable or name not in scope:
            dv = default_value(f)
            if dv is not None:
                scope[name] = dv

        cur = scope.get(name)
        if f.computed is not None:
            tasks.append((f, scope, path))
        elif f.object is not None and f.object.schema is not None and isinstance(cur, dict):
            _seed(e, f.object.schema, cur, path, tasks, root, errs)
        elif f.list is not None and len(f.list.items) >= 1 and isinstance(cur, list):
            for i, el in enumerate(cur):
                it = list_item_def(f.list, i)
                if (
                    it is not None
                    and it.object is not None
                    and it.object.schema is not None
                    and isinstance(el, dict)
                ):
                    _seed(e, it.object.schema, el, f"{path}[{i}]", tasks, root, errs)
        elif f.map is not None and f.map.value_schema is not None and isinstance(cur, dict):
            for k, el in cur.items():
                if isinstance(el, dict):
                    _seed(e, f.map.value_schema, el, join_path(path, k), tasks, root, errs)
        elif f.one_of is not None:
            sel = select_variant(f.one_of, cur)
            if sel is not None:
                _seed(e, sel[0], sel[1], path, tasks, root, errs)
        elif f.ref is not None:
            def_ = e.schema.defs.get(ref_def_key(f.ref))
            if def_ is not None and isinstance(cur, dict):
                _seed(e, def_, cur, path, tasks, root, errs)


def _run_normalize(
    e: Engine,
    schema: Schema,
    scope: NativeStruct,
    root: NativeStruct,
    errs: list[ValidationError],
) -> None:
    for f in schema.fields:
        name = f.name
        cur = scope.get(name)
        if cur is None:
            continue
        if not field_is_active(e, f, root, name, None):
            continue
        norm = f.normalize or ""
        if norm != "":
            value, err = e.eval(norm, {"this": cur, "root": root})
            if err is not None:
                errs.append(expr_err(name, norm, f"normalize: {err}"))
            else:
                scope[name] = value
                cur = value
        if f.object is not None and f.object.schema is not None and isinstance(cur, dict):
            _run_normalize(e, f.object.schema, cur, root, errs)
        elif f.list is not None and len(f.list.items) >= 1 and isinstance(cur, list):
            for i, el in enumerate(cur):
                it = list_item_def(f.list, i)
                if (
                    it is not None
                    and it.object is not None
                    and it.object.schema is not None
                    and isinstance(el, dict)
                ):
                    _run_normalize(e, it.object.schema, el, root, errs)
        elif f.map is not None and f.map.value_schema is not None and isinstance(cur, dict):
            for el in cur.values():
                if isinstance(el, dict):
                    _run_normalize(e, f.map.value_schema, el, root, errs)
        elif f.one_of is not None:
            sel = select_variant(f.one_of, cur)
            if sel is not None:
                _run_normalize(e, sel[0], sel[1], root, errs)
        elif f.ref is not None:
            def_ = e.schema.defs.get(ref_def_key(f.ref))
            if def_ is not None and isinstance(cur, dict):
                _run_normalize(e, def_, cur, root, errs)


def _run_compute(
    e: Engine,
    root: NativeStruct,
    tasks: list[tuple[SchemaField, NativeStruct, str]],
    errs: list[ValidationError],
) -> None:
    if not tasks:
        return
    by_path = {path: (f, scope) for f, scope, path in tasks}
    deps: dict[str, list[str]] = {}
    for f, _scope, path in tasks:
        assert f.computed is not None  # noqa: S101 - collected as computed
        deps[path] = [d for d in e.expr_deps(f.computed.expr) if d != path and d in by_path]

    color: dict[str, int] = {}
    order: list[str] = []

    def visit(p: str) -> bool:
        c = color.get(p, 0)
        if c == 1:
            return False
        if c == 2:
            return True
        color[p] = 1
        for d in deps.get(p, []):
            if not visit(d):
                return False
        color[p] = 2
        order.append(p)
        return True

    for _f, _scope, path in tasks:
        if color.get(path) != 2 and not visit(path):
            errs.append(schema_err(path, "computed field cycle"))

    for path in order:
        f, scope = by_path[path]
        assert f.computed is not None  # noqa: S101
        value, err = e.eval(f.computed.expr, {"root": root})
        if err is not None:
            errs.append(expr_err(path, f.computed.expr, f"compute: {err}"))
            continue
        shaped = shape_result(f.computed.result, value)
        if shaped is _MISMATCH:
            errs.append(
                expr_err(path, f.computed.expr, "compute: result does not match declared type")
            )
            continue
        scope[f.name] = shaped


_MISMATCH = object()


def shape_result(rt: SchemaFieldResultType | None, x: Native) -> Native:
    if x is None:
        return None
    match rt:
        case None | SchemaFieldResultType.UNSPECIFIED | SchemaFieldResultType.JSON:
            return x
        case SchemaFieldResultType.DOUBLE:
            n = as_float(x)
            return n if n is not None else _MISMATCH  # type: ignore[return-value]
        case SchemaFieldResultType.INT64:
            n = as_int(x)
            return n if n is not None else _MISMATCH  # type: ignore[return-value]
        case SchemaFieldResultType.UINT64:
            n = as_uint(x)
            return n if n is not None else _MISMATCH  # type: ignore[return-value]
        case SchemaFieldResultType.BOOL:
            return x if isinstance(x, bool) else _MISMATCH  # type: ignore[return-value]
        case SchemaFieldResultType.STRING:
            return x if isinstance(x, str) else _MISMATCH  # type: ignore[return-value]
        case SchemaFieldResultType.DURATION:
            return x if isinstance(x, dt.timedelta) else _MISMATCH  # type: ignore[return-value]
        case SchemaFieldResultType.TIMESTAMP:
            return x if isinstance(x, dt.datetime) else _MISMATCH  # type: ignore[return-value]
        case SchemaFieldResultType.BYTES:
            return x if isinstance(x, bytes) else _MISMATCH  # type: ignore[return-value]
        case _:
            return _MISMATCH  # type: ignore[return-value]


def coerce_input(f: SchemaField, val: Native) -> Native | None:
    if not isinstance(val, str):
        return None
    if f.int32 is not None or f.int64 is not None:
        try:
            return int(val, 10)
        except ValueError:
            return None
    if f.uint32 is not None or f.uint64 is not None:
        if val.startswith("-"):
            return None
        try:
            return int(val, 10)
        except ValueError:
            return None
    if f.float is not None or f.double is not None:
        try:
            return float(val)
        except ValueError:
            return None
    if f.bool is not None:
        return True if val == "true" else False if val == "false" else None
    if f.bytes is not None:
        try:
            return base64.b64decode(val, validate=True)
        except binascii.Error:
            return None
    if f.duration is not None:
        return parse_go_duration(val)
    if f.timestamp is not None:
        return parse_rfc3339(val)
    return None


def default_value(f: SchemaField) -> Native | None:
    if f.float is not None and f.float.default is not None:
        return cast("Native", f.float.default)
    if f.double is not None and f.double.default is not None:
        return cast("Native", f.double.default)
    if f.int32 is not None and f.int32.default is not None:
        return cast("Native", f.int32.default)
    if f.int64 is not None and f.int64.default is not None:
        return cast("Native", f.int64.default)
    if f.uint32 is not None and f.uint32.default is not None:
        return cast("Native", f.uint32.default)
    if f.uint64 is not None and f.uint64.default is not None:
        return cast("Native", f.uint64.default)
    if f.bool is not None and f.bool.default is not None:
        return cast("Native", f.bool.default)
    if f.string is not None and f.string.default is not None:
        return cast("Native", f.string.default)
    if f.bytes is not None and f.bytes.default:
        return cast("Native", f.bytes.default)
    if f.choice is not None and f.choice.default is not None:
        return to_native(f.choice.default)
    if f.duration is not None and f.duration.default is not None:
        return cast("Native", f.duration.default)
    if f.timestamp is not None and f.timestamp.default is not None:
        return cast("Native", f.timestamp.default)
    if f.json is not None and f.json.default is not None:
        return to_native(f.json.default)
    return None


def choice_options(e: Engine, name: str, root: NativeStruct) -> list[Native] | None:
    f = next((x for x in e.schema.fields if x.name == name), None)
    if f is None or f.choice is None:
        return None
    src = f.choice.options_expr or ""
    if src == "":
        return [to_native(o.value) for o in f.choice.options]
    value, err = e.eval(src, {"root": root})
    return value if err is None and isinstance(value, list) else None


def list_count(e: Engine, name: str, root: NativeStruct) -> int | None:
    f = next((x for x in e.schema.fields if x.name == name), None)
    if f is None or f.list is None:
        return None
    ce = f.list.count_expr or ""
    if ce == "":
        return None
    value, err = e.eval(ce, {"root": root})
    if err is not None:
        return None
    n = as_int(value)
    return n if n is not None and n >= 0 else None
